### ISSUE TITLE
No newline at end of .xsh file causes failure

### ISSUE THREAD

**Issue by mackeyja92** [xonsh/xonsh#1517]
Not having a new line at the end of my `.xsh` file causes all sorts of weird errors to be thrown. Here is an example when trying to do `bash-source nvm.sh` 

```
Traceback (most recent call last):
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 10323, in wrapped_simple_command
    r = f(args, i)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 14407, in source_alias
    builtins.execx(fp.read(), 'exec', builtins.__xonsh_ctx__)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 16152, in exec
    transform=transform)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 16113, in compile
    tree = self.parse(input, ctx, mode=mode, transform=transform)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 16084, in parse
    tree, input = self._parse_ctx_free(input, mode=mode)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/__amalgam__.py", line 16166, in _parse_ctx_free
    debug_level=(self.debug_level > 1))
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/parsers/base.py", line 307, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/vendor/lib/python3.5/site-packages/ply/yacc.py", line 331, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/vendor/lib/python3.5/site-packages/ply/yacc.py", line 1181, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/vendor/lib/python3.5/site-packages/ply/yacc.py", line 193, in call_errorfunc
    r = errorfunc(token)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/parsers/base.py", line 2340, in p_error
    self.xonsh_code)
  File "/usr/local/Cellar/xonsh/0.4.3/libexec/lib/python3.5/site-packages/xonsh/parsers/base.py", line 404, in _parse_error
    err_line = line.splitlines()[loc.lineno - 1]
IndexError: list index out of range
```

**Comment by scopatz** [comment]
#1517 might not be a complete fix for this, but it will at least allow the error to be reported correctly.

**Comment by mackeyja92** [comment]
```
+------------------+----------+
| xonsh            | 0.4.3    |
| Python           | 3.5.1    |
| PLY              | 3.7      |
| have readline    | True     |
| prompt toolkit   | None     |
| shell type       | readline |
| pygments         | None     |
| on posix         | True     |
| on linux         | False    |
| on darwin        | True     |
| on windows       | False    |
| on cygwin        | False    |
| is superuser     | False    |
| default encoding | utf-8    |
| git SHA          | None     |
+------------------+----------+
```

**Comment by laerus** [comment]
closing this since #1518 has been merged, feel free to reopen!

**Comment by laerus** [comment]
my bad, this is not the complete fix, reopening

**Comment by scopatz** [comment]
Actually, I can no longer reproduce this error.  Can anyone else?

**Comment by laerus** [comment]
i was never able to reproduce this,

btw is there a reason why on `get_code` method of the `XonshImportHook`
there is this line `src = src if src.endswith('\n') else src + '\n'` ??

**Comment by scopatz** [comment]
This is basically the reason.  Python's exec mode / our parser implementation requires that the code end in a newline.

**Comment by gforsyth** [comment]
I'm closing this.  Please reopen if the issue persists!

**Comment by flavius** [comment]
The bug can be reproduced with: https://gist.github.com/flavius/6d63648c1439714a77341cbc9c7c8f4d

Another bug I've found and which could be related can be reproduced with a similar code (without the import): https://gist.github.com/flavius/9eec3070781f175c37e465e24e45cad5

**Comment by scopatz** [comment]
@flavius - thanks for the hint! Can you please try the fix in #1760?

**Comment by gforsyth** [comment]
Fixed in #1760

**Comment by flavius** [comment]
@scopatz @gforsyth I've checked out 2a0f778ca454cde82576b66908d63d7f8cce151c and it works! Thanks again!
